### PR TITLE
Asus Zenfone 2 Laser (720p): Add Display,Backlight and Touchscreen support.

### PR DESCRIPTION
**Pull request by antony-jr**
This PR adds support for display panel, backlight and touchscreen for Asus Zenfone 2 (720p) (Codename: asus-z00l).

### CODE REVIEW

**Comment by antony-jr** [comment]
Oops too many style errors. Please review only the logic, I will take care of the style.

**Review by Minecrell — CHANGES_REQUESTED** [review]

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts [hunk 1]
```diff
@@ -48,10 +50,60 @@
...
 &blsp1_uart2 {
 	status = "okay";
 };
 
+&dsi0 {
+	pinctrl-names = "default", "sleep";
+        pinctrl-0 = <&mdss_default>;
+        pinctrl-1 = <&mdss_sleep>;
+        panel@0 {
+                compatible = "mdss,otm1284a";
```
[the remark this answers is not in the document]
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:85]: So should I write a documentation in ```Documentation/devicetree/...``` too? like how asus-z00t does? FYI asus-z00t is already in mainline kernel. But only the dts file is missing.

[the remark this answers is not in the document]
Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:85]: You don't need to write device tree documentation for merge in this repository, but you would need to write it when upstreaming your patches.

File: drivers/gpu/drm/panel/panel-asus-z00l-otm1284a.c [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	ret = mipi_dsi_dcs_set_display_brightness(dsi, brightness);
+	if (ret < 0)
+		return ret;
+
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
+	return 0;
+}
+
+static int otm1284a_bl_get_brightness(struct backlight_device *bl)
```
[the remark this answers is not in the document]
Comment (antony-jr) [reply drivers/gpu/drm/panel/panel-asus-z00l-otm1284a.c]: Yes there was. I thought actual_brightness gave the right value but it only gives -61 all the time so I removed the function.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts [hunk 3]
```diff
@@ -48,10 +50,59 @@
...
+&dsi0 {
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&mdss_default>;
+	pinctrl-1 = <&mdss_sleep>;
+	panel@0 {
+		compatible = "asus,z00l-otm1284a";
+		reg = <0>;
+
+		reset-gpios = <&msmgpio 25 GPIO_ACTIVE_LOW>;
+		backlight-gpios = <&msmgpio 21 GPIO_ACTIVE_LOW>;
```
Comment (antony-jr) [inline arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:89]: I have to make note here, Removing this line(```backlight-gpios```) has no effect. But it is in downstream.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts [hunk 4]
```diff
@@ -48,10 +50,59 @@
...
+
+		pinctrl-0 = <&lcd_bl_en_default>;
```
Comment (antony-jr) [inline arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:91]: This line also has no effect.

Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:91]: pinctrl-0 needs to be paired with pinctrl-names to have effect.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 3 above)
Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:89]: You need to generate your panel driver with `--backlight-gpio` to make use of this.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 4 above)
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:91]: Let me try with that.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 3 above)
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:89]: Okay I used the driver with --backlight-gpio and set backlight-gpios, Then the display stopped working. But this time when I press the power key I can see the Phosh lockscreen for a second and it turns off. So I think this needs to be avoided.

Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:89]: Have you tried GPIO_ACTIVE_HIGH for the backlight?

Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: Yep. GPIO_ACTIVE_HIGH fixed it.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 4 above)
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:92]: Okay I've given ```pinctrl-names = "default"``` but no effect from this, The display is working fine. I'm just curious what this adds to the kernel? I still see the glitch when changing brightness but it is stable when using it display.

**Comment by antony-jr** [comment]
@Minecrell Anything else that needs to be changed?

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: You didn't implement it like I suggested, you should put `asus,z00l-panel` here and let lk2nd replace it with `asus,z00l-otm1284a` after detecting your panel :)

**Comment by Minecrell** [comment]
See above.

Also, I think we should generate the panel driver with a power supply/regulator (`--regulator power`).

Downstream source suggests this should be `pm8916_l8`: https://github.com/msm8916-mainline/android_kernel_qcom_msm8916/blob/1a445c1767d9a60f4cce9fca61c9e4c84e2b44c3/arch/arm/boot/dts/qcom/msm8916-mdss.dtsi#L123

i.e. `power-supply = <&pm8916_l8>`.

**Comment by antony-jr** [comment]
> See above.
> 
> Also, I think we should generate the panel driver with a power supply/regulator (`--regulator power`).
> 
> Downstream source suggests this should be `pm8916_l8`: https://github.com/msm8916-mainline/android_kernel_qcom_msm8916/blob/1a445c1767d9a60f4cce9fca61c9e4c84e2b44c3/arch/arm/boot/dts/qcom/msm8916-mdss.dtsi#L123
> 
> i.e. `power-supply = <&pm8916_l8>`.

Thanks for the review. I will look into this ASAP.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 1 above)
[the remark this answers is not in the document]
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: After using asus,z00l-panel in the dts file. Now the display does not work. I can note that the panel section in lk2nd is showing ```qcom,mdss_dsi_otm1284a_720p_video```.

[the remark this answers is not in the document]
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: @Minecrell it fails to register the DSI host -> ```msm_dsi_manager_register: failed to register mipi dsi host 
for DSI 0``` when using ```asus,z00l-panel```

[the remark this answers is not in the document]
Comment (Minecrell) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: Did you build + install the new lk2nd? https://github.com/msm8916-mainline/lk2nd/pull/41

[the remark this answers is not in the document]
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: I did build from the master branch just yesterday and installed the new one. I will try using the panel-selection branch and report back. Thanks for your time.

[the remark this answers is not in the document]
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts]: I can confirm this works with using lk2nd/panel-selection branch.

**Comment by antony-jr** [comment]
> See above.
> 
> Also, I think we should generate the panel driver with a power supply/regulator (`--regulator power`).
> 
> Downstream source suggests this should be `pm8916_l8`: https://github.com/msm8916-mainline/android_kernel_qcom_msm8916/blob/1a445c1767d9a60f4cce9fca61c9e4c84e2b44c3/arch/arm/boot/dts/qcom/msm8916-mdss.dtsi#L123
> 
> i.e. `power-supply = <&pm8916_l8>`.

Okay I've done the requested changes. Also generated the driver with power supply and given ```power-supply = <&pm8916_l8>;```.

**Comment by Minecrell** [comment]
Changes look good, but can you make a separate commit for panel driver and the dts part? i.e.

- `drm/panel: ...`
- `arm64: dts: qcom: ...`

Since those will be applied separately later if you decide to upstream them.

Personally I would prefer if touchscreen was also a separate commit, but this isn't strictly required.

**Comment by antony-jr** [comment]
> Changes look good, but can you make a separate commit for panel driver and the dts part? i.e.
> 
>     * `drm/panel: ...`
> 
>     * `arm64: dts: qcom: ...`
> 
> 
> Since those will be applied separately later if you decide to upstream them.
> 
> Personally I would prefer if touchscreen was also a separate commit, but this isn't strictly required.

Okay. I will do this in a moment.

**Comment by antony-jr** [comment]
Okay made changes in separate commits.

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts [hunk 5]
```diff
@@ -185,6 +238,47 @@
 		bias-pull-up;
 	};
 
+	lcd_bl_en_default: lcd-bl-en-default {
+		pins = "gpio97";
```
Comment (Minecrell) [inline arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:242]: I'm really sorry that I keep finding things... Why is this `gpio97` rather than gpio21? :)

**Review by Minecrell — CHANGES_REQUESTED** [review]

File: arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts (hunk 5 above)
Comment (antony-jr) [reply arch/arm64/boot/dts/qcom/msm8916-asus-z00l.dts:242]: I think I made a mistake here. I read a comment. https://github.com/LineageOS/android_kernel_asus_msm8916/blob/1a45c63742b8c3253a38c2ff97b672918c88d8df/arch/arm/boot/dts/qcom/msm8916-mtp-ze550kl.dtsi#L491

Reviews are always welcomed.

**Review by Minecrell — APPROVED** [review]

**Comment by Minecrell** [comment]
Pushed, thanks :)

**Pull request closed without merging**